lib/pbench/agent/tool_data_sink.py

```diff
@@ -1295,19 +1300,45 @@ def execute(self):
                 )
             self.logger.debug("published %s", self._to_client_channel)
 
-            for data in self._from_client_chan.fetch_json(self.logger):
-                ret_val = self._valid_data(data)
-                if ret_val is None:
+            # Begin listening for client state signals, and react/respond
+            # accordingly upon signal receipt
+            for signal in self.sig_resp.listen():
+                client = signal.publisher_id
+                action = signal.event
+                self.logger.debug(
+                    "client %r, action %r, metadata %r", client, action, signal.metadata
+                )
+                if action == "shutdown":
+                    # This simply means that a SignalExporter/Client we
+                    # were listening to has shut down, no action needed
+                    self.logger.debug("Client %s has shut down", client)
                     continue
-                action, directory, args = ret_val[0], ret_val[1], ret_val[2]
-                self.execute_action(action, directory, args, data)
-            self.logger.info("Tool Data Sink terminating")
+                if action == "initialization":
+                    self.sig_resp.srespond(signal)
+                    continue
+                if not signal.metadata:
+                    self.logger.warning(
+                        "Signal for %s received, but required metadata not included",
+                        action,
+                    )
+                    continue
+                if not self._is_valid_data(action, signal.metadata):
+                    continue
+                data = signal.metadata
+                data["action"] = action
+                self.execute_action(client, data)
+                if action == "terminate":
+                    self.logger.info("Tool Data Sink terminating")
+                    self._send_client_status(client, action, "success")
+                    break
```

Comment:
What changed with state signals that we need to explicitly use a `break` statement here?

Reply:
Used to exit the listen() for loop

Reply:
Right but the previous `for` loop did not need a break because in `execute_action()` the "terminate" action closed the `_from_client_chan`.  Can the same be done so that the generator returned by `.listen()` will exit gracefully?